Migrate AvatarIcon

thebinij opened the pull request:
## Explanation
* Closes #18883  

## Screenshots/Screencaps

<!-- If you're making a change to the UI, make sure to capture a screenshot or a short video showing off your work! -->

### Before

<!-- How did the UI you changed look before your changes? Drag your file(s) below this line: -->

https://github.com/MetaMask/metamask-extension/assets/93417120/4d790c96-fc33-44a6-89fb-b74e603ec47b


### After

<!-- How does it look now? Drag your file(s) below this line: -->

https://github.com/MetaMask/metamask-extension/assets/93417120/e290348f-5194-4a03-bb92-b976e0176f31



## Manual Testing Steps

No functional changes

<!--
How should reviewers and QA manually test your changes? For instance:

- Go to this screen
- Do this
- Then do this
-->

## Pre-merge author checklist

- [x] I've clearly explained:
  - [x] What problem this PR is solving
- [ ] Sufficient automated test coverage has been added

## Pre-merge reviewer checklist

- [ ] Manual testing (e.g. pull and build branch, run in browser, test code being changed)
- [x] PR is linked to the appropriate GitHub issue
- [ ] **IF** this PR fixes a bug in the release milestone, add this PR to the release milestone

If further QA is required (e.g. new feature, complex testing steps, large refactor), add the `Extension QA Board` label.

In this case, a QA Engineer approval will be be required.

thebinij committed:
# This is a combination of 2 commits.
# This is the 1st commit message:

Migrate AvatarIcon

# This is the commit message #2:

fixing lint

thebinij commented:
Hi @georgewrmarshall, Please review this PR. Thanks!

georgewrmarshall commented on ui/components/component-library/avatar-icon/avatar-icon.types.ts:
```diff
@@ -0,0 +1,60 @@
+  Xl = 'xl',
+}
+
+export interface AvatarIconProps extends BoxProps {
```
Can we extend `AvatarBaseProps` here?

thebinij replied:
```javascript
export interface AvatarIconProps extends AvatarBaseProps{
 ```
TS complains:
- Interface 'AvatarIconProps' incorrectly extends interface 'AvatarBaseProps'.
  Types of property 'size' are incompatible
-   Types of property 'color' are incompatible.

#### If we extend both
```javascript
export interface AvatarIconProps extends BoxProps, AvatarBaseProps{
 ```
TS complains:
- Named property 'children' of types 'BoxProps' and AvatarBaseProps  are not identical.

- Named property 'textAlign' of types 'BoxProps' and 'AvatarBaseProps' are not identical.

georgewrmarshall reviewed (commented):
Looking good. Left a suggestion

thebinij commented on ui/components/component-library/avatar-icon/avatar-icon.tsx:
```diff
@@ -0,0 +1,55 @@
+        <Icon
+          color={IconColor.inherit}
+          name={iconName}
+          size={size as unknown as IconSize}
```
hey @georgewrmarshall, what do you think about this? I have done this to avoid TS error since 'size' of IconProps only accepts IconSize.

georgewrmarshall replied:
Hmm I see. Could we create a mapping of the AvatarIconSize to the IconSize enum somehow?

```
const avatarSizeToIconSizeMap: Record<AvatarIconSize, IconSize> = {
  [AvatarIconSize.Xs]: IconSize.Xs,
  [AvatarIconSize.Sm]: IconSize.Sm,
  [AvatarIconSize.Md]: IconSize.Md,
  [AvatarIconSize.Lg]: IconSize.Lg,
  [AvatarIconSize.Xl]: IconSize.Xl,
};

const AvatarIcon: React.FC<AvatarIconProps> = ({ size }) => {
  const iconSize = avatarSizeToIconSizeMap[size]; // Map AvatarIconSize to IconSize

  return <Icon size={iconSize} />;
};
```

thebinij commented on ui/components/component-library/avatar-icon/avatar-icon.types.ts:
```diff
@@ -0,0 +1,35 @@
+   * Defaults to AvatarIconSize.Md
+   */
+  size?: AvatarIconSize;
+  /**
+   * The color of the text inside the AvatarIcon
+   * Defaults to TextColor.primaryDefault
+   */
+  color?: TextColor | IconColor;
```
Hi @georgewrmarshall, Should I add union of TextColor and IconColor to 'color' of AvatarBaseProps ? Or should I just remove all instances of IconColor?

georgewrmarshall replied:
Hmm this is because for most Avatars when we think of color we are thinking of the color of the text but here it is for the icon. I think it would be ok to add the union of TextColor and IconColor to `color` of AvatarBaseProps

georgewrmarshall commented on ui/components/component-library/avatar-icon/avatar-icon.types.ts:
```diff
@@ -0,0 +1,35 @@
+  Xl = 'xl',
+}
+
+export interface AvatarIconProps extends Omit<AvatarBaseProps, 'color'> {
```
Can we also omit `children` so we don't have to pass `undefined` as a prop. The `<Icon />` component should suffice as the children prop

thebinij replied:
OKay, I will do that.

georgewrmarshall reviewed (commented):
Left one suggestion

georgewrmarshall reviewed (commented):
Left a suggestion about mapping the size prop

georgewrmarshall commented:
Hey @thebinij, Thanks for your time and effort in creating this pull request. As you're probably aware we have identified an issue related to the typing of the `Box` component, specifically with the polymorphic `as` prop. To address this issue, we have an open ticket at https://github.com/MetaMask/metamask-extension/issues/19239 and a draft pull request at https://github.com/MetaMask/metamask-extension/pull/19363 that needs to be merged before we can proceed with your PR. Thanks for your patience

thebinij commented:
Hi @georgewrmarshall, Is there anything I can update in the PR before so you approve to merge it?

garrettbear committed:
Migrate AvatarIcon

fixing lint

fixing size issue

IconProps extends BaseProps instead of BoxProps

Omit children from IconProps

fixing lint issues

mapping avatarIconSize to Iconsize

replace deprecated

garrettbear committed:
AvatarIcon cleanup

georgewrmarshall committed:
Small update to allow storybook iconName control to work

georgewrmarshall reviewed (approved):
LGTM! Made a tiny change to make the storybook controls to work as intended for the `iconName` prop. Thanks for your contribution @thebinij and support @garrettbear

georgewrmarshall commented on ui/components/component-library/avatar-icon/avatar-icon.stories.tsx:
```diff
@@ -46,11 +46,12 @@ export default {
 const Template: StoryFn<typeof AvatarIcon> = (args) => {
-  return <AvatarIcon {...args} iconName={IconName.SwapHorizontal} />;
+  return <AvatarIcon {...args} />;
 };
```
## After


https://github.com/MetaMask/metamask-extension/assets/8112138/a70ec83f-e36c-42ba-8974-dbc0a4c86be6
